### DISCUSSION TITLE
Discussion on `Should`, `Must` and the RFC-2119

### DISCUSSION THREAD

**Discussion by macrogreg** [fluentassertions/fluentassertions#1923] [General]
Apologies for the duplication. I submitted an issue on this (https://github.com/fluentassertions/fluentassertions/issues/1922) and then resized that discussions are active for this project, so re-posting here. Not sure what the maintainers consider the more appropriate place, so sorry for the dup.

---

### Background

This is not a bug, but rather a question / suggestion for the future. :)
Fluent Assertions is a brilliant framework. I am glad that it exists and I am grateful to the authors.

There is one thing about it that bugs me a lot. It is likely just due to my being overly compulsive about naming things in a deliberate and correct manner, yet I would be very interested and grateful to to learn whether there are others who feel similar. :)

My question is about a central concept of _Fluent Assertions_: the `Should(..)` method.

_Fluent Assertions_ does an amazing job making assertions read easily. Yet, a test case is also a "specification" of how the tested code is expected to behave. And, there exists a de-factor standard that describes guidelines for the English words to use in requirements specifications: the infamous [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).

Based on that, what _Fluent Assertions_ expresses using `Should(..)`, MUST really me expressed using the verb "must".

For those who are not familiar, the [aforementioned RFC](https://www.ietf.org/rfc/rfc2119.txt) is extremely short, check it out. The relevant part is:
* MUST: This word, or the terms "REQUIRED" or "SHALL", mean that the definition is an absolute requirement of the specification.
* SHOULD: This word, or the adjective "RECOMMENDED", mean that there may exist valid reasons in particular circumstances to ignore a particular item, but the full implications must be understood and carefully weighed before choosing a different course.

### Question 1: History

I am curious to learn the authors' thoughts on this topic. Did the choice `Should(..)` vs. `Must(..)` come up explicitly in the early days of _Fluent Assertions_? If yes, what were the reasons for the choice taken? Or was it the case that by the time this came up, there was already so much code in the wild based on `Should(..)` that a change could not be realistically considered for compatibility reasons? Or, perhaps, I am the only one worried about it? :)

### Question 2: Future / Adjustments

Assume that after learning some of the background in question 1 I wanted to use RFC-2119-style working in my own assertions, i.e. I wanted to replace `Should(..)` with `Must(..)`. Assume also that I wanted to share the solution for doing that with the community in case other people like that idea and want to apply it in a way that does not disrupt _Fluent Assertions_ at its core. 
In the context of the fact that I am in no way an exert on the internals of Fluent Assertions, would the authors (or any of the experts) have some guidance on the implementation strategy?

Based on a first looks, it seems that it may be possible to duplicate the [AssertionExtensions](https://github.com/fluentassertions/fluentassertions/blob/c33006c2e8857f7a02f1acffbc1e8a45ebd8853b/Src/FluentAssertions/AssertionExtensions.cs#L35-L1085) class and to define a `Must(..)` equivalent for every of the `Should(..)` overloads found there.

If I did so,
 - would it be preferable to just call through to the corresponding `Should(..)` version,
 - or to use the same implementation body as the respective `Should(..)` without actually calling into it?

Would such an approach break anything around the stack walks done by the framework?
Are there any other gotchas I should expect? 
Would there be any interest in including such an _optional_ extension in some way into _Fluent Assertions_ directly? Perhaps tucked away into a sub-namespace or in some other non-disruptive manner?
If this whole topic a crazy idea? :)

### Conclusion

I realize that people are busy. I respect your time and because of that I am ever more grateful for any feedback I might get from 
_Fluent Assertions_ experts on this topic.

Thank you.

**Comment by dennisdoomen** [comment]
Thank you for raising this discussion. I actually remembered that I regretted my choice to go for `should` many years ago. I'm quite open to the idea of switching to `must` in the next major release, depending on the feedback from the community.

**Reply by dagilleland** [reply]
Reminds me of the [regret Tony Hoare had over creating the Null Reference concept](https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare/).

**Comment by IT-VBFK** [comment]
Same here :)

If a test only _should_ be true, why are we testing ;)

A test IMHO have to be a particular scenario which **have to** pass... otherwise this test is useless

**Comment by dennisdoomen** [comment]
https://twitter.com/ddoomen/status/1522295222478249984

**Comment by jnyrup** [comment]
Q1: 
I can't comment much on the _early_ history since the project was already ~6 years old when I joined.
Had we created the library today `Must` might well have been the better choice, it's also two characters shorter than `Should`.

I'm often conservative about changes (my comments across issues and PRs will confirm), but renaming `Should` to `Must` is probably _the_ most breaking change I can think of.
To list some things it would break:
* ABI
* API
* Several nuget packages extending Fluent Assertions
* Documentation
* StackOverflow questions and answers
* Tutorials and examples all over the internet
* Every assertion written in all code bases consuming FA
* (My muscle memory)

I don't think the change outweights what it breaks.

Q2:
I haven't tried out making a clone of `Should` named `Must` that calls the constructors directly, but our code expects `.Should(` to be present in the call stack and I _think_ it might also have some expectation that the subject comes right before `.Should(`.

https://github.com/fluentassertions/fluentassertions/blob/c33006c2e8857f7a02f1acffbc1e8a45ebd8853b/Src/FluentAssertions/CallerIdentification/ShouldCallParsingStrategy.cs#L7

If you intend to proceed with making extension methods, I'm much open to see if we can adjust/extend the caller identification code to support using `Must` (or possibly any other word) in addition to `Should`.

**Reply by macrogreg** [reply]
Thanks, folks, for the discussion.
I agree on the heavy impact of such a change, and that if it would be considered at all, it would need to be offered to the public in a way that carefully manages the breaking changes.

My wild guess is that during the vote on twitter, the breaking changes were a driving factor.
Although, personally, I'd go with `var.Please().BeTrue();` 😄  

I have a very full schedule for May, but I suggest that in June I spend some time on trying to make the `.Must(..)` extension methods work. Once I am working on it, I'll ask for some pointers for what needs to be tweaked except for the extension methods themselves. If the required changes are reasonable and can be made to the system without too much churn, then we can make them and offer the "Must" package in any way you prefer: completely separately (independent extension), as a separate, optional package of FluentAssertions, or as an optional feature within FluentAssertions.

Does that sound reasonable from your perspective?

**Reply by jnyrup** [reply]
Take your time to experiment as it's not blocking anyone. 
Share your findings and let's then discuss what the best path forward might be.

**Reply by DanStevens** [reply]
I find this topic interesting as I am often writing [characterization tests](https://en.wikipedia.org/wiki/Characterization_test) for legacy code, in which case the term 'Should' is somewhat unfitting because I'm writing to tests to describe what the code does as-is without respect to correctness. The more fitting term here would be 'Is', but unfortunately that doesn't fit well with FluentAssertions since `LegacyMethod().Is.Be(true)` is no longer proper English. `LegacyMethod().Is(true)` would be nice but I imagine that would be a bigger change than 'Should' to 'Must'.  I suppose 'Appears To' (e.g. `AppearsTo.Be(true)`) would fit the syntax better while still being fluent, although it does sound a bit contrived.

**Reply by dennisdoomen** [reply]
It should be relatively easy to support `Must` next to `Should`

**Comment by antoinebj** [comment]
Any update on this?
I think it would be good to transition from `Should()` to `Must()` to match the semantic clarity of RFCs. In an upcoming major version we could have an equivalent `Must()` for each `Should()` overload, and mark the latter as obsolete with a message guiding to use `Must()`. For those who have warnings as errors, it's an easy conversion of code, just a replace all instances of the `.Should()` string to `.Must()` in the unit test projects. It should (must? :D) be done in 30 seconds. And in a later major version, `Should()` can be removed.

**Comment by dennisdoomen** [comment]
Would be nice to have some kind of code generator generating overloads for every `Should`.

**Comment by jnyrup** [comment]
> Any update on this?

What kind of update are you seeking?

I wrote in https://github.com/fluentassertions/fluentassertions/discussions/1923#discussioncomment-2740725 why I don't think changing this today pulls its own weight.

If I saw a library that had two methods `Should()` and `Must()` I would assume there was a difference between them, since why would there otherwise be two methods instead of one, and what that difference was.

**Reply by antoinebj** [reply]
My mistake, I had not understood from that comment that the suggestion was dropped altogether. Also the discussion does not have a "Closed" state, so that added to my misunderstanding.

> If I saw a library that had two methods Should() and Must() I would assume there was a difference between them, since why would there otherwise be two methods instead of one, and what that difference was.

Would you not see `Should` stricken through in the IDE autocomplete suggestions if it was marked with the `ObsoleteAttribute`, plus some warnings? Would that not avoid any significant confusion of that sort?

**Reply by jnyrup** [reply]
I'm sorry, I just realized my latest comment had a quite hostile tone, that wasn't my intention.

I vote for closing this topic as won't do.

**Reply by dennisdoomen** [reply]
I didn't realize we already closed this topic.
